internal/api/server.go

```diff
@@ -443,37 +452,78 @@ func (s *Server) Auth(c echo.Context) error {
+		return c.JSON(http.StatusBadRequest, map[string]string{
+			"error": "Invalid request format",
+		})
+	}
+
+	// Validate required fields
+	if err := clientutil.ValidateAuthRequest(
+		req.Message, req.Signature, req.PublicKey, req.DerivePath, req.ChainCodeHex,
+	); err != nil {
+		return c.JSON(http.StatusBadRequest, map[string]string{
+			"error": err.Error(),
+		})
+	}
+
+	// Verify the message format matches the expected format from client
+	expectedMessage := clientutil.GenerateHexMessage(req.PublicKey)
+	if req.Message != expectedMessage {
+		s.logger.Warnf("Message mismatch: expected %s, got %s", expectedMessage, req.Message)
+		return c.JSON(http.StatusUnauthorized, map[string]string{
+			"error": "Message content mismatch",
+		})
 	}
 
-	msgBytes, err := hex.DecodeString(strings.TrimPrefix(req.Message, "0x"))
+	// Decode message from hex (remove 0x prefix first)
+	msgWithoutPrefix := strings.TrimPrefix(req.Message, "0x")
+	msgBytes, err := hex.DecodeString(msgWithoutPrefix)
 	if err != nil {
 		s.logger.Errorf("failed to decode message: %v", err)
-		return c.NoContent(http.StatusBadRequest)
+		return c.JSON(http.StatusBadRequest, map[string]string{
+			"error": "Invalid message format",
+		})
 	}
 
-	sigBytes, err := hex.DecodeString(strings.TrimPrefix(req.Signature, "0x"))
+	// Decode signature from hex (remove 0x prefix first)
+	sigWithoutPrefix := strings.TrimPrefix(req.Signature, "0x")
+	sigBytes, err := hex.DecodeString(sigWithoutPrefix)
 	if err != nil {
 		s.logger.Errorf("failed to decode signature: %v", err)
-		return c.NoContent(http.StatusBadRequest)
+		return c.JSON(http.StatusBadRequest, map[string]string{
+			"error": "Invalid signature format",
+		})
 	}
 
-	success, err := sigutil.VerifySignature(req.PublicKey,
-		req.ChainCodeHex,
-		msgBytes,
-		sigBytes)
+	// Verify the signature using our utility
+	success, err := sigutil.VerifySignature(req.PublicKey, req.ChainCodeHex, req.DerivePath, msgBytes, sigBytes)
 	if err != nil {
 		s.logger.Errorf("signature verification failed: %v", err)
-		return c.NoContent(http.StatusUnauthorized)
+		return c.JSON(http.StatusUnauthorized, map[string]string{
+			"error": "Signature verification failed: " + err.Error(),
+		})
 	}
 	if !success {
-		return c.NoContent(http.StatusUnauthorized)
+		return c.JSON(http.StatusUnauthorized, map[string]string{
+			"error": "Invalid signature",
+		})
 	}
 
-	token, err := s.authService.GenerateToken()
+	// Generate JWT token with the public key
+	token, err := s.authService.GenerateToken(req.PublicKey)
 	if err != nil {
 		s.logger.Error("failed to generate token:", err)
-		return c.NoContent(http.StatusInternalServerError)
+		return c.JSON(http.StatusInternalServerError, map[string]string{
+			"error": "Failed to generate auth token",
+		})
+	}
+
+	// Store logged-in user's public key in cache for quick access
+	cacheKey := "user_pubkey:" + token[0:10]                                          // Use first part of token as a cache key
```

Comment:
why token[0:10]? why not the whole key?

Reply:
its enough for uniqueness and does not compromise security (replay attacks). should i switch to full token?

Reply:
if you are comfortable with it , that's fine